### ISSUE TITLE
Latest master @f75f7c4 causes crash to storenodes

### ISSUE THREAD

**Issue by sbc64** [status-im/nim-waku#577]
The latest commit https://github.com/status-im/nim-waku/commit/f75f7c4751727016e478e9eee640cb406e171d4b is causing the nodes I run to crash. I have two relay + store nodes peered with each other using both the staticnode and storenode flags pointing at each other.

The crash happens to the second node that is being started. So it doesn't matter which node it is, the one that crashes is always the second one to start. I turned on and off some of the command line flags and the culprit seems to be `--storenode` flag that is causing. I reverted to the previous commit https://github.com/status-im/nim-waku/commit/131ce6c43b64be846e34948094d27bc8e4b6dc68 and nodes work as expected.

```
DBG 2021-05-27 12:11:51.675+01:00 UPnP                                       topics="nat" tid=5598 file=nat.nim:67 msg="Internet Gateway Device not found. Giving up."
DBG 2021-05-27 12:11:51.676+01:00 NAT-PMP                                    topics="nat" tid=5598 file=nat.nim:91 msg="the gateway does not support nat-pmp"
INF 2021-05-27 12:11:51.682+01:00 Initializing networking                    topics="wakunode" tid=5598 file=wakunode2.nim:126 hostAddress=/ip4/0.0.0.0/tcp/60001 announcedAddresses=@[]
DBG 2021-05-27 12:11:51.682+01:00 About to accept incoming connection        topics="libp2p switch" tid=5598 file=switch.nim:318
DBG 2021-05-27 12:11:51.682+01:00 Started libp2p node                        topics="libp2p switch" tid=5598 file=switch.nim:368 peer="(peerId: \"16Uiu2HAkxthSUou6qdD5igzt6E4niDRbsXGTU6GpKNGKRahv6TTX\", addrs: @[\"/ip4/0.0.0.0/tcp/60001\"], protocols: @[\"/ipfs/id/1.0.0\"], protoVersion: \"\", agentVersion: \"\")"
INF 2021-05-27 12:11:51.682+01:00 PeerInfo                                   topics="wakunode" tid=5598 file=wakunode2.nim:160 peerId=16U*hv6TTX addrs=@[/ip4/0.0.0.0/tcp/60001]
INF 2021-05-27 12:11:51.682+01:00 Listening on                               topics="wakunode" tid=5598 file=wakunode2.nim:163 full=/ip4/0.0.0.0/tcp/60001/p2p/16Uiu2HAkxthSUou6qdD5igzt6E4niDRbsXGTU6GpKNGKRahv6TTX
INF 2021-05-27 12:11:51.682+01:00 mounting store                             topics="wakunode" tid=5598 file=wakunode2.nim:396
DBG 2021-05-27 12:11:51.682+01:00 mounting store without swap                topics="wakunode" tid=5598 file=wakunode2.nim:399
DBG 2021-05-27 12:11:51.682+01:00 init                                       topics="wakunode" tid=5598 file=waku_store.nim:392
INF 2021-05-27 12:11:51.682+01:00 dialPeer                                   topics="wakunode" tid=5598 file=wakunode2.nim:537 address=/ip4/127.0.0.1/tcp/60002/p2p/16Uiu2HAmVishq9LSgkSKrUF6TLEpCb6Zm1r2vaLnnBfBuLCwcZ3N
DBG 2021-05-27 12:11:51.682+01:00 Adding peer to manager                     topics="wakupeers" tid=5598 file=peer_manager.nim:170 peerId=16U*CwcZ3N addr=/ip4/127.0.0.1/tcp/60002 proto=/vac/waku/store/2.0.0-beta3
DBG 2021-05-27 12:11:51.683+01:00 resume                                     topics="wakustore" tid=5598 file=waku_store.nim:521 currentEpochTime=1622113911.683043
DBG 2021-05-27 12:11:51.683+01:00 the  offline time window is                topics="wakustore" tid=5598 file=waku_store.nim:527 lastSeenTime=0.0 currentTime=1622313911.683043
DBG 2021-05-27 12:11:51.683+01:00 no candidate list is provided, selecting a random peer topics="wakustore" tid=5598 file=waku_store.nim:551
DBG 2021-05-27 12:11:51.683+01:00 a peer is selected from peer manager       topics="wakustore" tid=5598 file=waku_store.nim:559
INF 2021-05-27 12:11:51.683+01:00 Dialing peer from manager                  topics="wakupeers" tid=5598 file=peer_manager.nim:47 wireAddr=/ip4/127.0.0.1/tcp/60002 peerId=16U*CwcZ3N
DBG 2021-05-27 12:11:51.683+01:00 Dialing peer                               topics="libp2p switch" tid=5598 file=switch.nim:102 peerId=16U*CwcZ3N
DBG 2021-05-27 12:11:51.686+01:00 no handlers                                topics="libp2p multistream" tid=5598 file=multistream.nim:168 conn=16U*CwcZ3N:60af7e77e913bd170c45cfd5 protocol=/vac/waku/relay/2.0.0-beta2
DBG 2021-05-27 12:11:51.687+01:00 decodeMsg: decoded message                 topics="libp2p identify" tid=5598 file=identify.nim:94 pubkey=secp25...28]))) addresses=@[/ip4/0.0.0.0/tcp/60002] protocols="@[\"/ipfs/id/1.0.0\", \"/vac/waku/store/2.0.0-beta3\", \"/vac/waku/relay/2.0.0-beta2\"]" observable_address=Some(/ip4/127.0.0.1/tcp/42746) proto_version="Some(\"ipfs/0.1.0\")" agent_version="Some(\"nim-libp2p/0.0.1\")"
DBG 2021-05-27 12:11:51.687+01:00 no handlers                                topics="libp2p multistream" tid=5598 file=multistream.nim:168 conn=16U*CwcZ3N:60af7e77e913bd170c45cfd5 protocol=/vac/waku/relay/2.0.0-beta2
DBG 2021-05-27 12:11:51.687+01:00 Dial successful                            topics="libp2p switch" tid=5598 file=switch.nim:154 conn=16*CwcZ3N:60af7e77e913bd170c45cfd2 peerInfo="(peerId: \"16Uiu2HAmVishq9LSgkSKrUF6TLEpCb6Zm1r2vaLnnBfBuLCwcZ3N\", addrs: @[\"/ip4/0.0.0.0/tcp/60002\"], protocols: @[\"/ipfs/id/1.0.0\", \"/vac/waku/store/2.0.0-beta3\", \"/vac/waku/relay/2.0.0-beta2\"], protoVersion: \"ipfs/0.1.0\", agentVersion: \"nim-libp2p/0.0.1\")"
/home/sebas/repos/walletconnect/nim-waku/waku/v2/node/wakunode2.nim(707) wakunode2
/home/sebas/repos/walletconnect/nim-waku/vendor/nim-chronos/chronos/asyncmacro2.nim(1092) waitFor
/home/sebas/repos/walletconnect/nim-waku/vendor/nim-chronos/chronos/asyncmacro2.nim(281) poll
/home/sebas/repos/walletconnect/nim-waku/vendor/nim-chronos/chronos/asyncloop.nim(74) colonanonymous
/home/sebas/repos/walletconnect/nim-waku/vendor/nim-chronos/chronos/asyncloop.nim(169) readExactly
[[reraised from:
/home/sebas/repos/walletconnect/nim-waku/waku/v2/node/wakunode2.nim(707) wakunode2
/home/sebas/repos/walletconnect/nim-waku/vendor/nim-chronos/chronos/asyncloop.nim(1092) waitFor
/home/sebas/repos/walletconnect/nim-waku/vendor/nim-chronos/chronos/asyncloop.nim(281) poll
/home/sebas/repos/walletconnect/nim-waku/vendor/nim-chronos/chronos/asyncmacro2.nim(74) colonanonymous
/home/sebas/repos/walletconnect/nim-waku/vendor/nim-chronos/chronos/asyncmacro2.nim(367) readVarint
]]
[[reraised from:
/home/sebas/repos/walletconnect/nim-waku/waku/v2/node/wakunode2.nim(707) wakunode2
/home/sebas/repos/walletconnect/nim-waku/vendor/nim-chronos/chronos/asyncloop.nim(1092) waitFor
/home/sebas/repos/walletconnect/nim-waku/vendor/nim-chronos/chronos/asyncloop.nim(281) poll
/home/sebas/repos/walletconnect/nim-waku/vendor/nim-chronos/chronos/asyncmacro2.nim(74) colonanonymous
/home/sebas/repos/walletconnect/nim-waku/vendor/nim-chronos/chronos/asyncmacro2.nim(367) readLp
]]
[[reraised from:
/home/sebas/repos/walletconnect/nim-waku/waku/v2/node/wakunode2.nim(707) wakunode2
/home/sebas/repos/walletconnect/nim-waku/vendor/nim-chronos/chronos/asyncloop.nim(1092) waitFor
/home/sebas/repos/walletconnect/nim-waku/vendor/nim-chronos/chronos/asyncloop.nim(281) poll
/home/sebas/repos/walletconnect/nim-waku/vendor/nim-chronos/chronos/asyncmacro2.nim(74) colonanonymous
/home/sebas/repos/walletconnect/nim-waku/vendor/nim-chronos/chronos/asyncmacro2.nim(367) queryFrom
]]
[[reraised from:
/home/sebas/repos/walletconnect/nim-waku/waku/v2/node/wakunode2.nim(707) wakunode2
/home/sebas/repos/walletconnect/nim-waku/vendor/nim-chronos/chronos/asyncloop.nim(1092) waitFor
/home/sebas/repos/walletconnect/nim-waku/vendor/nim-chronos/chronos/asyncloop.nim(281) poll
/home/sebas/repos/walletconnect/nim-waku/vendor/nim-chronos/chronos/asyncmacro2.nim(74) colonanonymous
/home/sebas/repos/walletconnect/nim-waku/vendor/nim-chronos/chronos/asyncmacro2.nim(367) resume
]]
[[reraised from:
/home/sebas/repos/walletconnect/nim-waku/waku/v2/node/wakunode2.nim(707) wakunode2
/home/sebas/repos/walletconnect/nim-waku/vendor/nim-chronos/chronos/asyncloop.nim(1092) waitFor
/home/sebas/repos/walletconnect/nim-waku/vendor/nim-chronos/chronos/asyncloop.nim(281) poll
/home/sebas/repos/walletconnect/nim-waku/vendor/nim-chronos/chronos/asyncmacro2.nim(74) colonanonymous
/home/sebas/repos/walletconnect/nim-waku/vendor/nim-chronos/chronos/asyncmacro2.nim(367) resume
]]
[[reraised from:
/home/sebas/repos/walletconnect/nim-waku/waku/v2/node/wakunode2.nim(707) wakunode2
]]
Error: unhandled exception: Stream EOF! [LPStreamEOFError]

```

Thanks for your support  :pray:

**Comment by oskarth** [comment]
cc @staheri14

**Comment by jm-clius** [comment]
The issue here is that `relay` is not yet mounted when `store` attempts to resume during [`wakunode2` startup procedures](https://github.com/status-im/nim-waku/blob/master/waku/v2/node/wakunode2.nim#L707) (and in order to dial you need to have `relay` mounted first). You could move the resume functionality till after you've mounted `relay`, but that would (currently) mean that you start relaying messages before you've properly resumed your history, which is not ideal. To delay starting `relay` until you've finished resuming history is also not an option as it stands, since this is tied in with starting the switch which you need to do before the `resume`.
In other words, our protocol mounting behaviour (passive i.t.o. network participation, e.g. initialise and mount protocol on `switch`) and protocol starting behaviour (often active, e.g. resume history, subscribe to default topics, etc.) needs to be untangled.

**Comment by staheri14** [comment]
Thanks for identifying the issue @sbc64!
Also thanks @jm-clius for your insights! you are right about the order issue. I am working on it.

**Comment by staheri14** [comment]
@jm-clius You have mentioned a great point!  However, I am wondering why this issue hasn't showed up in this test
https://github.com/status-im/nim-waku/blob/f75f7c4751727016e478e9eee640cb406e171d4b/tests/v2/test_wakunode.nim#L690

**Comment by staheri14** [comment]
Moved the resume to after the relay is mounted, and things are working just fine.
@jm-clius While resuming history after the start of relay protocol is not ideal but is not problematic either, wdyt? In my opinion, old messages that are resumed after the start of the relay protocol will end up having (receiver generated) timestamps that are higher than the more recent messages, but that is not really an issue given that this can also happen due to the network delay.

**Comment by staheri14** [comment]
> in order to dial you need to have `relay` mounted first

@jm-clius Could you please elaborate on why `relay` is required? Can't a node directly dial another node without mounting the `relay` protocol?

**Comment by oskarth** [comment]
> In other words, our protocol mounting behaviour (passive i.t.o. network participation, e.g. initialise and mount protocol on switch) and protocol starting behaviour (often active, e.g. resume history, subscribe to default topics, etc.) needs to be untangled.

Do we have an issue to track this?

**Comment by jm-clius** [comment]
> @jm-clius Could you please elaborate on why `relay` is required? Can't a node directly dial another node without mounting the `relay` protocol?

In short, it is required _if the node you're dialing also has `relay` mounted_ (which explains why the unit test at https://github.com/status-im/nim-waku/blob/f75f7c4751727016e478e9eee640cb406e171d4b/tests/v2/test_wakunode.nim#L690 is working).

Explainer in the Background Recap for this PR: https://github.com/status-im/nim-waku/pull/529

**Comment by jm-clius** [comment]
> Do we have an issue to track this?

We do now: https://github.com/status-im/nim-waku/issues/581 😃

**Comment by jm-clius** [comment]
@staheri14 I think this can now be closed?

Comment from @sbc64 on Telegram:
> https://github.com/status-im/nim-waku/commit/8f741de483a0dd4ac180dc634191b1060ebd2737 is working  💯
